config: Allow specifying helper dirs with $BINDIR as base directory

filbranden opened the pull request:
This should make it easier to locate helper binaries relative to where the main binary was installed, which should be useful in installations such as Homebrew which install packages under a versioned directory.

Use a `$BINDIR` magic token as a prefix in the helper path to indicate it should be relative to the directory where the binary is located. This is somewhat familiar to the syntax used in the shell and Makefile and is still quite explicit about the behavior (as opposed to, say, making all relative paths be relative to the directory of the binary.)

Update `podman` config on Darwin to look for helpers such as `gvproxy` under `$BINDIR/../libexec/podman`, which is the ultimate objective of this code change.

Tested: Updated vendored package in podman, built it with `podman-remote`, copied `gvproxy` to a `libexec/podman` at the same level as `bin/podman` and confirmed that `podman machine start` worked as expected.

Fixes: containers/podman#12161

Related: PR #946 (cc @ashley-cui)

Signed-off-by: Filipe Brandenburger <filbranden@gmail.com>

<!--- Please read the [contributing guidelines](https://github.com/containers/common-files/blob/master/.github/CONTRIBUTING.md) before proceeding --->

vrothberg reviewed (commented):
@ashley-cui @baude PTAL

rhatdan commented on pkg/config/config.go:
```diff
@@ -1245,13 +1247,35 @@ func (c *Config) ActiveDestination() (uri, identity string, err error) {
+		if path == bindirPrefix || strings.HasPrefix(path, bindirPrefix+string(filepath.Separator)) {
+			// Calculate the path to the executable first time we encounter a $BINDIR prefix.
+			if bindirPath == "" && bindirErr == nil {
+				podman, bindirErr := os.Executable()
```
nit: This should not be called podman, since this library is used for other tools like buildah.

Just call it executable

filbranden replied:
Sure that makes sense, I'm using `execPath` now.

(It's a bit odd though, since `config_darwin.go` in this same package uses `libexec/podman` for the helper directory, but I guess that's perhaps expected since the helpers are somewhat shared between both?)

rhatdan commented:
Thanks @filbranden

ashley-cui commented on pkg/config/config.go:
```diff
@@ -1261,6 +1285,9 @@ func (c *Config) FindHelperBinary(name string, searchPATH bool) (string, error)
 	}
 	configHint := "To resolve this error, set the helper_binaries_dir key in the `[engine]` section of containers.conf to the directory containing your helper binaries."
+	if bindirErr != nil {
+		configHint = fmt.Sprintf("Error looking up path relative to %s: [%s]  %s", bindirPrefix, bindirErr, configHint)
```
Should this be a logrus.Warnf instead of a print?

filbranden replied:
Oh yes indeed logging is much nicer!

By not having to track the error result, I ended up refactoring the code considerably, moving `bindirPath` calculation to a standalone function.

In the new code, I also track whether `os.Executable()` (or resolving symlinks from it) ever failed, in order to only log about it once.

I hope you still like how the new code looks, let me know if you'd like me to make any changes to it.

ashley-cui commented:
Thanks @filbranden ! A nit from Dan and a nit from me, but other than that, LGTM!

filbranden commented:
Updated the PR.

One case I had previously missed was that the called binary is actually a symbolic link, now I added a call to `filepath.EvalSymlinks` to cover this case. I tested this with a symbolic link now and confirmed that this works. (This is the actual Homebrew setup, so I'm glad I caught this now before we merge it.)

Also split it into two separate commits, one with the feature to support the `$BINDIR` magic, and the second one updating the Darwin config to use the feature to look for helpers in a path relative to the binary itself.

PTAL. Thanks!

rhatdan commented:
LGTM

rhatdan commented:
/approve
/lgtm
